### ISSUE TITLE
CurlHandler completes SendAsync task even if request content not fully sent

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#9006]
This is a feature or a bug depending on who you talk to.

CurlHandler completes the Task returned from SendAsync when all of the response headers have been received.  That's by design.  However, it does this regardless of whether all of the request content has been received.  While rare, it's possible a server might send the response headers and even some or all of the body before the request content has been sent/received, e.g. in a streaming scenario where it was echoing/transforming/etc. the request content out as response content.

WinHttpHandler and the implementation on desktop complete the SendAsync task only once all of the request content has been sent and all of the response headers have been received.

The CurlHandler design is potentially problematic, however, as HttpClient explicitly disposes of the request content once the SendAsync task completes.  If that task completes and the request content is still being sent, obvious problems can result.

Several potential solutions:
- Modify HttpClient to stop Dispose'ing of the request content, e.g. https://github.com/dotnet/corefx/issues/1794
- Moving the responsibility for Dispose'ing of the HttpContent to the HttpClientHandler, which can know when it's safe to dispose, e.g. https://github.com/dotnet/corefx/pull/8894
- Modifying CurlHandler to only complete the SendAsync task once the request content has all been sent, e.g. https://github.com/dotnet/corefx/pull/8884

**Comment by Priya91** [comment]
@stephentoub The linked potential solution PRs are merged, is there work remaining for this issue?

**Comment by stephentoub** [comment]
The underlying problem is still there.  The linked PRs were examples of how it could be addressed, but they were closed rather than merged.  (I would not be surprised if an investigation into https://github.com/dotnet/corefx/issues/16259 shows that this issue is the cause.)

**Comment by Priya91** [comment]
Thanks for the info, will keep the issue active and tracked for 2.0.

**Comment by stephentoub** [comment]
@davidsh, @geoffkizer, @cipop, @Priya91, I think we need to make one of these two changes, and do so for 2.0:
1) Modify HttpClient to stop Dispose'ing of the request content, e.g. #1794
2) Moving the responsibility for Dispose'ing of the HttpContent to the HttpClientHandler, which can know when it's safe to dispose, e.g. #8894.

My preference is for (1).  The downside to (1) is that existing code that relies on the request content being disposed will now not get the prompt disposal; if disposal matters for it, it will either have the disposal delayed until finalization kicks in, or it will need to be disposed by the developer.  I'd venture to guess that the vast majority of usage won't be affected by this, e.g. where there either is no request content or the content doesn't actually need disposal, e.g. a ByteArrayContent, a StringContent, a StreamContent with a Stream where disposal is basically a nop (e.g. MemoryStream), etc.  The upside though is big: developers can now cache request objects and reuse them over and over (on serialized requests), a handler that supports duplex communication (e.g. CurlHandler) can now be used for that, etc.

(2) will mitigate the concerns around the content not being disposed, as the HttpClientHandler can do it when the request is entirely completed.  And it would still allow duplex communication for handlers that support it.  The downsides for (2) are that it wouldn't allow reuse of request objects, and every handler would need to be modified to do the disposal (but in reality, 99.99% of usage would be handled by the handlers we ship and we can modify those accordingly).

We're seeing more and more folks hit this issue on Unix as they communicate with servers that send their responses before the request is fully processed, and it's becoming a real issue we need to fix.  If there are other solutions beyond the two I've mentioned above that achieve the same goals, I'd be interested in hearing them, but those are the two that jump out at me.

**Comment by krwq** [comment]
@davidsh @geoffkizer @CIPop @Priya91 is anyone of you looking at this? Could you at least comment on approaches suggested by @stephentoub so someone with less background can pick it up if you do not have spare cycles?

**Comment by davidsh** [comment]
I think choice 1 is the cleanest choice. I think doing something like 2 is messy. I do worry though that since Desktop won't have this change developers might be confused when they write 'netstandard' based code. Eventually, we might need to make a similar breaking change in Desktop and include an opt-out mechanism (registry and/or appsettings config) like we have done in the past for breaking changes.

**Comment by Priya91** [comment]
@stephentoub I agree with 1 as well, and exposing a property on HttpContent for controlling the behavior sounds good. The problem in the issue exists with CurlHandler and libcurl, but the proposed fix differs .net core from the behavior on Desktop, so until we can have an API that can control this behavior, should we go with option 3 in the description, instead of disposing, make CurlHandler to wait SendAsync till the full request is sent out?

**Comment by karelz** [comment]
@stephentoub can you please help?

**Comment by stephentoub** [comment]
> should we go with option 3 in the description, instead of disposing, make CurlHandler to wait SendAsync till the full request is sent out?

I've spent some time today looking into this, and I can't find a way to do it well.  If we knew that libcurl would always request the request content and read it to its completion, then we could do this, but there are circumstances where it won't read the whole request content or may not read any of it, in which case if we block waiting for it to, we effectively deadlock.  We can play whack-a-mole trying to handle all of the cases we're aware of, but at the end of the day such interactions aren't documented and could change, and getting it wrong leads to very bad situations.

I also looked into (2) a bit, but that has its own non-trivial downsides.  Today code can use an HttpClientHandler without HttpClient, e.g. by using HttpMessageInvoker instead of HttpClient, in which case it bypasses this disposal logic (and other logic provided by the relatively thin HttpClient wrapper).  If we were to move the disposal logic to the HttpClientHandler, now such code that was never having its request content disposed will start to, which is bad and exactly the opposite of what we want.  It also causes problems for chains of handlers, e.g. if HttpClient is given a handler that wraps CurlHandler, and that handler expects to be able to look at the request content after the SendAsync Task from CurlHandler completes but before it itself completes its SendAsync Task.  And while such a scenario may seem far fetched, we have it in the bits today: the diagnostic handler that's used to pass details off to DiagnosticListeners is injected in the middle like this, and it does pass off the request content after the wrapped SendAsync Task completes... if a listener was expecting a non-disposed request content at that point, it would be sorely disappointed.

So, while there is the possibility of a break due to (1), I think it's the least risky of the options. I'll submit a PR for that.
